vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,647 @@
+ *  |            |                     |          |
+ *  |            |                     |          |
+ *  +------------+                     +----------+
+ *
+ *
+ * External cross-reference
+ * ------------------------
+ * The only cross-reference to outside data structures is the cost_index
+ * corresponding to the data structure t_rr_index_data
+ * Details can be found in the definition of t_rr_index_data
+ *
+ * +---------+  cost_index   +------------------------------+
+ * | RRGraph |-------------->| cost_indexe_data[cost_index] | 
+ * +---------+               +------------------------------+
+ *
+ * Access to a node/edge/switch/segment, please use the StrongId created
+ * ---------------------------------------------------------------------
+ *    For node, use RRNodeId
+ *    For edge, use RREdgeId
+ *    For switch, use RRSwitchId
+ *    For segment, use RRSegmentId
+ *
+ *    These are the unique identifier for each data type.
+ *    To check if your id is valid or not, use the INVALID() function of StrongId class.
+ *    Example:
+ *       if (node_id == RRNodeId::INVALID()) {
+ *       }  
+ *
+ ***********************************************************************/
+#ifndef RR_GRAPH_OBJ_H
+#define RR_GRAPH_OBJ_H
+
+/*
+ * Notes in include header files in a head file 
+ * Only include the neccessary header files 
+ * that is required by the data types in the function/class declarations!
+ */
+/* Header files should be included in a sequence */
+/* Standard header files required go first */
+#include <limits>
+#include <vector>
+
+/* EXTERNAL library header files go second*/
+#include "vtr_vector.h"
+#include "vtr_range.h"
+#include "vtr_geometry.h"
+#include "arch_types.h"
+
+/* VPR header files go second*/
+#include "vpr_types.h"
+#include "rr_graph_fwd.h"
+
+class RRGraph {
+  public: /* Types */
+    /* Iterators used to create iterator-based loop for nodes/edges/switches/segments */
+    typedef vtr::vector<RRNodeId, RRNodeId>::const_iterator node_iterator;
+    typedef vtr::vector<RREdgeId, RREdgeId>::const_iterator edge_iterator;
+    typedef vtr::vector<RRSwitchId, RRSwitchId>::const_iterator switch_iterator;
+    typedef vtr::vector<RRSegmentId, RRSegmentId>::const_iterator segment_iterator;
+
+    /* Ranges used to create range-based loop for nodes/edges/switches/segments */
+    typedef vtr::Range<node_iterator> node_range;
+    typedef vtr::Range<edge_iterator> edge_range;
+    typedef vtr::Range<switch_iterator> switch_range;
+    typedef vtr::Range<segment_iterator> segment_range;
+
+  public: /* Accessors */
+    /* Aggregates: create range-based loops for nodes/edges/switches/segments
+     * To iterate over the nodes/edges/switches/segments in a RRGraph, 
+     *    using a range-based loop is suggested.
+     *  -----------------------------------------------------------------
+     *    Example: iterate over all the nodes
+     *      for (const RRNodeId& node : nodes()) {
```

Comment:
I think it would be helpful to declare an rr_graph here, and then use it in the example

RRGraph rr_graph;
for (const RRNodeId& node : rr_graph.nodes() )